packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,209 @@
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only be executed via
+    ///         cross-chain call from the other messenger OR if the message was already received once and is currently
+    ///         being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _source      Chain ID of the source chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _message     Message payload to call target with.
+    function relayMessage(
+        uint256 _destination,
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message
+    )
+        external
+        payable
+        nonReentrant
+    {
+        if (msg.sender != Predeploys.CROSS_L2_INBOX) revert RelayMessageCallerNotCrossL2Inbox();
+        if (CrossL2Inbox(Predeploys.CROSS_L2_INBOX).origin() != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+        }
+        if (_destination != block.chainid) revert MessageDestinationNotRelayChain();
+        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
+        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert MessageTargetL2ToL2CrossDomainMessenger();
+        }
+
+        bytes32 messageHash = keccak256(abi.encode(_destination, _source, _nonce, _sender, _target, _message));
+        if (successfulMessages[messageHash]) {
+            revert MessageAlreadyRelayed();
+        }
+
+        _storeMessageMetadata(_source, _sender);
+
+        successfulMessages[messageHash] = true;
+
+        bool success = _callWithAllGas(_target, _message);
+
+        if (success) {
+            emit RelayedMessage(messageHash);
+        } else {
+            successfulMessages[messageHash] = false;
+            emit FailedRelayedMessage(messageHash);
+        }
+    }
+
+    /// @notice Retrieves the next message nonce. Message version will be added to the upper two bytes of the message
+    ///         nonce. Message version allows us to treat messages as having different structures.
+    /// @return Nonce of the next message to be sent, with added message version.
+    function messageNonce() public view returns (uint256) {
+        return Encoding.encodeVersionedNonce(msgNonce, messageVersion);
+    }
+
+    /// @notice Stores message data such as sender and source in transient storage.
+    /// @param _source Chain ID of the source chain.
+    /// @param _sender Address of the sender of the message.
+    function _storeMessageMetadata(uint256 _source, address _sender) internal {
+        // Store non-zero value in slot for `entered`
+        TransientContext.set(ENTERED_SLOT, 1);
+
```

Comment:
We want to have a store then clear pattern. Since we are non reentrant, we want to store the value and then clear it at the end of the call frame. We don't need to hash together the values like inside the implementation of `set`